### PR TITLE
move sync_project_binding to controller manager

### PR DESCRIPTION
**Pull request by zreigz**
**What this PR does / why we need it**: move sync_project_binding to controller manager

```release-note
NONE
```

### CODE REVIEW

File: api/pkg/controller/rbac/rbac_controller.go [hunk 1]
```diff
@@ -311,24 +301,6 @@ func (c *Controller) enqueueProjectResource(obj interface{}, gvr schema.GroupVer
...
-	}
-	defer c.projectBindingsQueue.Done(key)
-
-	err := c.syncProjectBindings(key.(string))
-
-	c.handleErr(err, key, c.projectQueue)
-	return true
-}
-
 func (c *Controller) enqueueProjectBinding(projectBinding *kubermaticv1.UserProjectBinding) {
```
Comment (p0lyn0mial) [inline api/pkg/controller/rbac/rbac_controller.go]: rm `enqueueProjectBinding` method too.

File: api/pkg/controller/rbac/rbac_controller.go [hunk 2]
```diff
@@ -178,15 +178,6 @@ func New(metrics *Metrics, allClusterProviders []*ClusterProvider) (*Controller,
 			},
 			kind: kubermaticv1.SSHKeyKind,
 		},
-
-		{
```
Comment (p0lyn0mial) [inline api/pkg/controller/rbac/rbac_controller.go]: bring it back, we still want to generate RBAC for `UserProjectBindingKind`

File: api/pkg/controller/rbac/rbac_controller.go [hunk 3]
```diff
@@ -311,24 +301,6 @@ func (c *Controller) enqueueProjectResource(obj interface{}, gvr schema.GroupVer
 	c.projectResourcesQueue.Add(item)
 }
 
-func (c *Controller) runProjectBindingsWorker() {
-	for c.processProjectBindingNextItem() {
-	}
-}
-
-func (c *Controller) processProjectBindingNextItem() bool {
```
Comment (p0lyn0mial) [inline api/pkg/controller/rbac/rbac_controller.go:319]: remove `projectBindingsInformer`  https://github.com/kubermatic/kubermatic/blob/master/api/pkg/controller/rbac/rbac_controller.go#L134

File: api/pkg/controller/rbac/rbac_controller.go [hunk 4]
```diff
@@ -311,24 +301,6 @@ func (c *Controller) enqueueProjectResource(obj interface{}, gvr schema.GroupVer
...
-	key, quit := c.projectBindingsQueue.Get()
-	if quit {
-		return false
-	}
-	defer c.projectBindingsQueue.Done(key)
```
Comment (p0lyn0mial) [inline api/pkg/controller/rbac/rbac_controller.go:324]: rm `projectBindingsQueue` too.

File: api/cmd/rbac-generator/main.go [hunk 5]
```diff
@@ -190,6 +193,33 @@ func main() {
...
+			glog.Info("registering components")
+			if err := kubermaticv1.AddToScheme(mgr.GetScheme()); err != nil {
+				return err
+			}
+
+			if err := rbaccontroller.Add(mgr); err != nil {
+				return err
+			}
+
+			go func() {
```
Comment (p0lyn0mial) [inline api/cmd/rbac-generator/main.go]: if `mgr.Start` fails we will never stop processing, wouldn't be better to move it to a separate `g.Add` function ?

File: api/cmd/rbac-generator/main.go [hunk 6]
```diff
@@ -190,6 +193,33 @@ func main() {
...
+			if err != nil {
+				return err
+			}
+
+			mgr, err := manager.New(cfg, manager.Options{})
+			if err != nil {
+				glog.Errorf("failed to start kubebuilder manager: %v", err)
+				return err
+			}
+			glog.Info("registering components")
```
Comment (p0lyn0mial) [inline api/cmd/rbac-generator/main.go]: this is too verbose, I'd remove it.

File: api/pkg/controller/rbac/sync_project_binding_controller.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+	kubermaticv1 "github.com/kubermatic/kubermatic/api/pkg/crd/kubermatic/v1"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const ControllerName = "kubermatic_sync_projectbinding_controller"
```
Comment (p0lyn0mial) [inline api/pkg/controller/rbac/sync_project_binding_controller.go]: make it private.

File: api/pkg/controller/rbac/sync_project_binding_controller.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+	"fmt"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	kubermaticv1 "github.com/kubermatic/kubermatic/api/pkg/crd/kubermatic/v1"
```
Comment (p0lyn0mial) [inline api/pkg/controller/rbac/sync_project_binding_controller.go:8]: group imports.

File: api/pkg/controller/rbac/sync_project_binding_controller.go [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+	// Watch for changes to ClusterRoles
+	err = c.Watch(&source.Kind{Type: &kubermaticv1.UserProjectBinding{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// reconcileRBAC reconciles Cluster Role and Cluster Role Binding objects
```
Comment (p0lyn0mial) [inline api/pkg/controller/rbac/sync_project_binding_controller.go]: update comment.

File: api/pkg/controller/rbac/sync_project_binding_controller.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+func Add(mgr manager.Manager) error {
+	r := &reconcileSyncProjectBinding{Client: mgr.GetClient(), ctx: context.TODO()}
+	// Create a new controller
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to ClusterRoles
```
Comment (p0lyn0mial) [inline api/pkg/controller/rbac/sync_project_binding_controller.go]: update comment.

File: api/pkg/controller/rbac/sync_project_binding_controller.go [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+// reconcileRBAC reconciles Cluster Role and Cluster Role Binding objects
+type reconcileSyncProjectBinding struct {
+	ctx context.Context
+	client.Client
+}
+
+func (r *reconcileSyncProjectBinding) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+
```
Comment (p0lyn0mial) [inline api/pkg/controller/rbac/sync_project_binding_controller.go:46]: won't review as this was moved from from the previous controller.

**Review by p0lyn0mial — CHANGES_REQUESTED** [review]

**Comment by zreigz** [comment]
@p0lyn0mial PTAL

**Comment by p0lyn0mial** [comment]
what about metrics ? should we monitor at least the number of running controllers ?

**Comment by zreigz** [comment]
good point, right now we can initialize Prometheus in controller manager. All useful metrics will be added automatically

**Comment by p0lyn0mial** [comment]
/lgtm
/approve

we are going to add metrics in a separate pull.

**Pull request merged**